### DISCUSSION TITLE
Rules feedback  ❤️

### DISCUSSION THREAD

**Discussion by tommhuth** [NoA-Ignite-dk/eslint-config#7] [Ideas]
So I think we are at a pretty good place linting wise, but after using it for a while now there are some issues I feel work less great then others and that it would be good to have some discussion around, so we make sure the rules are helpful and don't reduce our productivity. 

- I get the idea behind **prefer default export**, but it's a problem when you have utility files where a single default export does make sense: it then forces you to either have more than one export, or refactor exports later as new util methods make their way into the file. 
- **Consistent returns** are nice, but the rule is currently not smart enough to understand `useEffect` (where we often do not have consistent returns for conditional effects)
-  **Leading decimals**: so does anyone really confuse these with property access?? Decimals is a core part of any animation work and the squiggly lines frankly just feel like noise 
- **Line length**: I'm all for sensible line-breaks, but too short a line length often makes code _less readable._The Canify repo is a great example of this.
- We currently use single tick **quotation marks**, but also allow double quotes in react properties anyway. Why not just pick double quotes everywhere and be able to truly enforce a single style. Also, don't think there is anything stopping us from allow single tick in react props anyway?
- We're enforcing **blank lines between class _properties_**? Are we sure about that? Seems a little excessive. 
- Force **direct return**, also seems a little unnecessary and often forces you to do pretty pointless refactors as the code evolves.  The linter can't possible know the future of the code and so should force any temp fixes for stuff that might be meaningless anyway.

**Comment by KasperAndersson** [comment]
In terms of **quotation marks**, I think `single quotes` makes sense in the context of _java-/typescript_, and `double quotes` in the context of _html/jsx_

**Reply by tommhuth** [reply]
yes i think that is the historical reason for that linting rule, but my point is that there isn't anything stopping us from picking _one_, and if people feel single quotes is off the table for jsx properties, then i suggest we pick double quotes for everything. right now we are kinda enforcing single quotes while at the same time not enforcing it consistently.

**Reply by jdpnielsen** [reply]
I personally don't mind single quotes in jsx, but i can see an argument for keeping it as double quotes for separation (although i don't know if this provides any kind of value). Eslint has an autofix for quotes, so converting code to single quotes is easy enough.

Anyway - we should configure the `jsx-quotes` rule to enforce consistency.

**Reply by KasperAndersson** [reply]
For me personally, `<button className="big-red-button" onClick={ () => onClickHandler('I hit the button') }>Don't click me</button>` is more readable than `<button className='big-red-button' onClick={ () => onClickHandler('I hit the button') }>Don't click me</button>`. Cause the `"` and `'` is in different context.

**Reply by tommhuth** [reply]
I totally disagree with that, also since you would never format it like that:

```tsx
<button 
    className="big-red-button"  
    onClick={ () => onClickHandler('I hit the button') }
>
    Don't click me
</button>
```

The readability here does not come from quote marks, but rather everything else.

**Reply by tommhuth** [reply]
also, do you really need to differentiate them? they are both totally strings

**Comment by jdpnielsen** [comment]
**prefer default export**: I agree, this should be disabled. I don't actually like default exports at all.

**Reply by tommhuth** [reply]
default exports are great for react components, but linting isn't going to be smart enough to enforce that in any way i think.

**Reply by jdpnielsen** [reply]
What is the benefit in react components?

**Reply by mthines** [reply]
`default export` for React Component also have the downside that it provides issues with Storybook.
That of course can be resolves in a later version, but right now you can't get the Props automatically generated from the TS types from a `default export`, but only `export const`

**Reply by jdpnielsen** [reply]
Rule disabled in v1.2.0: https://github.com/NoA-Ignite-dk/eslint-config/blob/master/CHANGELOG.md#120-2022-02-09

**Comment by jdpnielsen** [comment]
**Consistent returns**:

Can this not just be resolved using `return undefined`?
![image](https://user-images.githubusercontent.com/8746698/152792056-c322a2dc-4e53-439e-a1b5-6ea0d02addd2.png)

**Reply by tommhuth** [reply]
yes, but now you have to litter your code with lots of `return undefined` for now reason, and i really think explicit returns in `useEffects` should mark actual cleanups, not be there to satisfy the linter.  it just adds noise and takes focus away from what the code is _actually_ doing.

**Reply by jdpnielsen** [reply]
How about something like this instead?
![image](https://user-images.githubusercontent.com/8746698/152806433-9851471b-5eba-4dba-8b3c-03a0d9d9459a.png)

I think this maps much better onto how effects should be used & encourages [separation of concern](https://reactjs.org/docs/hooks-effect.html#tip-use-multiple-effects-to-separate-concerns).

**Reply by mthines** [reply]
I agree with @tommhuth. 
For regular functions the rule is great, but for `useEffect` it's clustering and adding noise.

If the return type is really important I tend to just add it as a type which then provides the error.
```tsx
function getString(name?: string): string {
  return name; // <- Eslint complain - `name` can be `undefined`
}

const getString = (name?: string): string => {
  return name; // <- Eslint complain - `name` can be `undefined`
}
```

**Reply by jdpnielsen** [reply]
> I agree with @tommhuth. For regular functions the rule is great, but for `useEffect` it's clustering and adding noise.
> 
> If the return type is really important I tend to just add it as a type which then provides the error.
> 
> ```tsx
> function getString(name?: string): string {
>   return name; // <- Eslint complain - `name` can be `undefined`
> }
> 
> const getString = (name?: string): string => {
>   return name; // <- Eslint complain - `name` can be `undefined`
> }
> ```

Just FYI we are discussing consistent returns - which is a lint rule which says that if there is one return statement in a function, all other paths should also end in a return (or a throw). This does not have anything to do with return types - of which i agree, it should not be required, as sometimes its better to let typescript infer the returned type.

**Reply by mthines** [reply]
I know, and I know the rule. I showed a bad example, sorry.
I personally agree with @tommhuth and I'm not a fan of the rule, as in many cases the codebase gets noisy, but I don't have strong feelings about this :)
I'm simply just sharing my thoughts. 

I do think the `return type` are relevant to this subject as it's basically what the `consistent-return` rule enforces for a function.

If I've defined a return type it will operate as if you've added a `return type`.

Therefore I think a better solution is to remove the rule, because you can always add the `return type` to a function, "marking" the  `return type` as important. 

For instance with the rule applied, it will complain if you have a `useEffect` which conditionally adds a timeout, when have a clean up.


### With the rule set to `error`.

#### `useEffect`
![image](https://user-images.githubusercontent.com/16097850/153171352-386c7c15-9f02-4aae-8b77-f9f2f8015e8f.png)

#### `Function` without `return type`
![image](https://user-images.githubusercontent.com/16097850/153173363-b4a89ed8-eb45-494e-a98b-0bb0bbe7f9e5.png)

### With the rule set to `off`.

#### `useEffect`
![image](https://user-images.githubusercontent.com/16097850/153171469-b4486da0-5209-4eec-b164-da759656f275.png)

#### `Function` with `return type` ⛔️
![image](https://user-images.githubusercontent.com/16097850/153172983-212045d8-4c9b-4ff6-ba68-ef22d16534f8.png)

#### `Function` with `return type` ✅
![image](https://user-images.githubusercontent.com/16097850/153172959-f300db96-8e71-4bf4-a58a-b87f26152e70.png)

#### `Function` without `return type`
![image](https://user-images.githubusercontent.com/16097850/153173736-cab8295b-ac23-40f1-bd61-d4ff99356ad0.png)

In my opinion this achieves the same, but removes the unnecessary noise for the functions where the `consistent-return` is not important.

Hope that clears out the confusion :)

**Reply by jdpnielsen** [reply]
That makes sense. I'm fine with removing the consistent return type rule

**Reply by jdpnielsen** [reply]
Rule disabled in v1.2.0: https://github.com/NoA-Ignite-dk/eslint-config/blob/master/CHANGELOG.md#120-2022-02-09

**Reply by tommhuth** [reply]
woo!

**Comment by jdpnielsen** [comment]
**Leading decimals**

@tommhuth I don't quite understand - can you maybe share a usage example in regards to animation?

![image](https://user-images.githubusercontent.com/8746698/152792332-70b6298f-ccac-4d3c-837c-1ceef8f45aac.png)

In my opinion `0.4` is much more readable than `.4` in the above example.

**Reply by tommhuth** [reply]
so per se i don't really care to much about the leading 0 when reading it, it's more when writing code, and skipping the leading 0 is so deeply ingrained in my style that i often end up thinking i have syntax errors before realize that i in fact, just forgot a leading 0. since the linter fixes this for you i'm ok with keeping it in, i'm justing venting a bit 😂

**Reply by mthines** [reply]
`0.4` ❤️

**Reply by KasperAndersson** [reply]
`onSave: eslint --fix` 🗡️

**Comment by jdpnielsen** [comment]
**blank lines between class properties**:
Agreed - i think we should use the `exceptAfterSingleLine` [option](https://eslint.org/docs/rules/lines-between-class-members) to allow grouped properties and singleline methods.

**Reply by tommhuth** [reply]
i'm ok with linebreak for methods, but between properties does not make sense to me

**Reply by jdpnielsen** [reply]
The rule should resolve this:
(`'@typescript-eslint/lines-between-class-members': ['error', 'always', { exceptAfterSingleLine: true }]`) 
![image](https://user-images.githubusercontent.com/8746698/152809143-aac5fa20-8f8e-49a2-974f-ebbcc60cda4c.png)

**Reply by jdpnielsen** [reply]
Rule enhanced in v1.2.0: https://github.com/NoA-Ignite-dk/eslint-config/blob/master/CHANGELOG.md#120-2022-02-09

**Comment by jdpnielsen** [comment]
**Line length**: The line length is currently configured to be max 140 chars. I think lines longer than this are problematic:

![image](https://user-images.githubusercontent.com/8746698/152793677-057459b5-6489-4f8d-ab95-965252c59874.png)

**Reply by tommhuth** [reply]
yes so actually it might be that canify uses some custom handling here, because 140 is pretty ok.  compare that to something like this from canify which really isn't helping code readability:
<img width="1009" alt="Screen Shot 2022-02-07 at 14 40 58" src="https://user-images.githubusercontent.com/1224941/152799092-1fcb9583-4d55-458d-b2c9-824dc38e2109.png">

**Reply by jdpnielsen** [reply]
I think that might be prettier, and not the eslint config

**Reply by mthines** [reply]
I'm not sure I think it's the 140 which makes the example hard to read.

I think a grouping of the variables, hooks and declarations also will make this more readable.
And potentially abstracting into reusable hooks which is grouped by context anyway, will also make it more readable.

**Comment by jdpnielsen** [comment]
**Direct return**: Which eslint rule is this? Do you have an example of the code which gives this lint error?

**Reply by tommhuth** [reply]
so direct returns for arrow functions forces stuff like this for arrays and objects:

```ts
// Not OK
const actors = useMemo(() => {
    return [ /* something */]
}, []);
// OK
const actors = useMemo(() => [
 	/* something */
], []);
```

it gets fiddly for several reason:
- the dev nows what is going to happen to this overtime, meaning that an explicit return might make sense in the future anyway
- an explicit return never makes the code unreadable, and also helps with keeping the line length  down 
- refactor from direct return to explicit return is fiddly (where do those squiggly guys go!?)

**Reply by jdpnielsen** [reply]
I agree, the lint rules should be more large here. But as a general FYI, there is a refactoring available for converting the arrow function:

![image](https://user-images.githubusercontent.com/8746698/152809800-2b2c6f7e-2cb7-425a-822a-da4ec3b2c48b.png)

**Reply by jdpnielsen** [reply]
Might have to add this, though: https://marketplace.visualstudio.com/items?itemName=nicoespeon.abracadabra

**Reply by mthines** [reply]
I love `Direct return` for presentational React Components that just receives props and render JSX, but I think we should value readability over code length, and some times for functions `Direct return` can make it less readable imo.

### React

```tsx
// 👍
const Component = ({ name }: Props) => (
  <div>
    <h2>{name}</h2>
  </div>
)
```

### Functions

```tsx
// 👍
const getName = (firstName: string, lastName?: string) => {
  if (lastName) {
    return `${firstName} ${lastName}`; 
  }

  return firstName;
}

// 👎
const getName = (firstName: string, lastName?: string) => lastName ? `${firstName} ${lastName}` : firstName;
```

**Reply by jdpnielsen** [reply]
Rule disabled in v1.2.0: https://github.com/NoA-Ignite-dk/eslint-config/blob/master/CHANGELOG.md#120-2022-02-09
